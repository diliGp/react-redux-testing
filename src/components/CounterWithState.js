import React, { Component } from 'react'

class CounterWithState extends Component {
    constructor(props) {
        super(props);

        this.state = {
            count: 0
        };
    }

    increment(amount) {
        this.setState({
            count: this.state.count + amount
        });
    }

    decrement(amount) {
        this.setState({
            count: this.state.count - amount
        });
    }

    render() {
        return (
            <div data-test="container-state-counter">
                <h1>Counter with State</h1>
                <section>
                    <h2>Current count: {this.state.count}</h2>
                    <div>
                        <button data-test="increment-btn" onClick={() => this.increment(1)}>+</button>
                        <button data-test="decrement-btn" onClick={() => this.decrement(1)}>-</button>
                    </div>
                </section>
            </div>
        )
    }
}

export default CounterWithState;
