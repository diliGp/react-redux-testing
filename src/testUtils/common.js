import { createStore, applyMiddleware } from 'redux';
import { rootReducer } from '../reducers';
import { middlewares } from '../store'

export const findByAttr = (wrapper, attrValue) => wrapper.find(`[data-test="${attrValue}"]`);

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
}