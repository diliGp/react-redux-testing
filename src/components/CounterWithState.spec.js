import React from 'react';
import { shallow } from "enzyme";
import CounterWithState from "./CounterWithState";
import { findByAttr } from '../testUtils/common';
import renderer from 'react-test-renderer';

describe('Counter with State', () => {
    let component;
    beforeEach(() => {
        component = shallow(<CounterWithState />);
    });

    it('Should capture snapshot to render without error', () => {
        const tree = renderer.create(<CounterWithState />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Should render without error', () => {
        const container = findByAttr(component, 'container-state-counter');
        expect(container.length).toBe(1);
    });

    it('Should be having initial count in state', () => {
        const initialState = component.state();
        expect(initialState.count).toBe(0);
    });

    it('Should contain incrementer', () => {
        const incrementer = findByAttr(component, 'increment-btn');
        expect(incrementer.length).toBe(1);
    });
    
    it('Should contain decrementer', () => {
        const decrementer = findByAttr(component, 'decrement-btn');
        expect(decrementer.length).toBe(1);
    });
    
    it('Should increase counter', () => {
        const incrementer = findByAttr(component, 'increment-btn');
        incrementer.simulate('click');
        const changedState = component.state();
        expect(changedState.count).toBe(1);
    });
    
    it('Should decrease counter', () => {
        const incrementer = findByAttr(component, 'decrement-btn');
        incrementer.simulate('click');
        const changedState = component.state();
        expect(changedState.count).toBe(-1);
    });

});