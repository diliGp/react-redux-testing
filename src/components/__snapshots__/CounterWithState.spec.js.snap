// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter with State Should capture snapshot to render without error 1`] = `
<div
  data-test="container-state-counter"
>
  <h1>
    Counter with State
  </h1>
  <section>
    <h2>
      Current count: 
      0
    </h2>
    <div>
      <button
        data-test="increment-btn"
        onClick={[Function]}
      >
        +
      </button>
      <button
        data-test="decrement-btn"
        onClick={[Function]}
      >
        -
      </button>
    </div>
  </section>
</div>
`;
